File: scripts/coverage-fraction.py
#!/usr/bin/env python3
"""The denominator for `seen_pcs` (devlog 0086).

Since 0078 the fuzz loop has reported a coverage number — 4789, then 5203, 5406, 5443 —
and every one of those was a bare count. A count is not a fraction: it says the number
went up, never how much is left. This script supplies the missing half.

KCOV records `_RET_IP_` of `__sanitizer_cov_trace_pc`, which is the address of the
instruction FOLLOWING the call. So the reachable coverage points are exactly the return
addresses of those calls, and they can be enumerated statically from vmlinux. Intersect
that set with what a run actually saw, attribute both sides to the object file each
function came from, and the result is a real percentage of the verification pass.

Usage: coverage-fraction.py <vmlinux> <fuzz log with PCSET lines> <kernel/bpf obj dir>
"""
import bisect
import collections
import re
import subprocess
import sys

# The verification pass, as this tree splits it: bpf_check() and everything it reaches to
# PROVE a program safe. core.o is the runtime, syscall.o the entry, and the map
# implementations are not the thing under test — they are reported but not counted.
VERIFICATION_PASS = {
    'verifier.o', 'states.o', 'liveness.o', 'cnum.o', 'tnum.o', 'backtrack.o',
    'const_fold.o', 'cfg.o', 'check_btf.o', 'fixups.o', 'log.o', 'disasm.o',
}

# Grouping for the never-entered remainder. The point of the whole measurement is to say
# where to aim next, and that answer is only legible in themes, not in 238 function names.
THEMES = [
    ('kfunc / BTF calls',          r'kfunc|btf_check_func|fetch_kfunc|pseudo_btf|ptr_to_btf'),
    ('log + disasm printing',      r'^print_|verifier_vlog|verbose|disasm|fmt_'),
    ('reference / spin lock',      r'reference|spin_lock|release_reg|rcu_protected|irq_flag'),
    ('attach target / trampoline', r'attach_target|tramp|struct_ops'),
    ('callbacks / subprogs',       r'callback|func_exit|subprog'),
    ('dynptr / iterator',          r'dynptr|_iter'),
    ('packet pointers',            r'pkt_pointer|pkt_ptr|find_good_pkt'),
    ('arena / map special',        r'arena|map_ptr|timer|wq|kptr|map_field'),
]

FUNC_RE = re.compile(r'^([0-9a-f]+) <(.+)>:$')
INSN_RE = re.compile(r'^\s*([0-9a-f]+):\t')


def static_sites(vmlinux):
    """addr -> function name, for every KCOV coverage point in the image."""
    sites, funcs, cur, pending = {}, [], None, False
    p = subprocess.Popen(['objdump', '-d', '--no-show-raw-insn', vmlinux],
                         stdout=subprocess.PIPE, text=True, bufsize=1 << 20)
    for line in p.stdout:
        m = FUNC_RE.match(line)
        if m:
            cur = m.group(2)
            funcs.append((int(m.group(1), 16), cur))
            pending = False
            continue
        m = INSN_RE.match(line)
        if not m:
            continue
        if pending:
            sites[int(m.group(1), 16)] = cur
            pending = False
        if 'call' in line and '__sanitizer_cov_trace_pc' in line:
            pending = True
    p.wait()
    funcs.sort()
    return sites, funcs


def symbol_owners(objdir):
    """function name -> object file, dropping every name more than one object defines.

    Every translation unit carries GCC's `_sub_I_65535_1`/`_sub_D_65535_0` stubs. Keying
    a dict by name hands them all to whichever file was read last, and in the first run
    of this script that single mistake put 5396 of 15241 coverage points — a third of the
    denominator — behind two symbols nothing could ever enter. An ambiguous name cannot
    attribute a PC, so it is dropped rather than guessed.
    """
    import glob
    import os
    owners = collections.defaultdict(set)
    for o in sorted(glob.glob(os.path.join(objdir, '*.o'))):
        base = os.path.basename(o)
        if base.endswith('.mod.o') or base == 'built-in.o':
            continue
        out = subprocess.run(['nm', '--defined-only', o], capture_output=True, text=True)
        for line in out.stdout.splitlines():
            f = line.split()
            if len(f) == 3 and f[1] in ('t', 'T'):
                owners[f[2]].add(base)
    return {s: next(iter(v)) for s, v in owners.items() if len(v) == 1}


def observed(log):
    pcs = set()
    for line in open(log, errors='replace'):
        if line.startswith('PCSET ') and 'total=' not in line:
            for t in line.split()[1:]:
                if re.fullmatch(r'[0-9a-f]{16}', t):
                    pcs.add(int(t, 16))
    return pcs


def main():
    vmlinux, log, objdir = sys.argv[1], sys.argv[2], sys.argv[3]
    sites, funcs = static_sites(vmlinux)
    sym2obj = symbol_owners(objdir)
    pcs = observed(log)
    hit = pcs & set(sites)

    def obj(fn):
        return sym2obj.get(fn) or sym2obj.get(fn.split('.')[0])

    print(f"COVFRAC image_sites={len(sites)} observed={len(pcs)} matched={len(hit)} "
          f"unmatched={len(pcs) - len(hit)}")
    # An unmatched PC is a coverage point this script's disassembly walk did not locate —
    # a jump table inside .text desynchronises objdump. They land in functions we DO
    # reach, and the same walk loses sites on the denominator side, so the error is not
    # one-sided. Reported so it is a measured limit and not a silent one.
    starts = [a for a, _ in funcs]
    lost = collections.Counter()
    for p in sorted(pcs - hit):
        i = bisect.bisect_right(starts, p) - 1
        lost[funcs[i][1] if i >= 0 else '?'] += 1
    print("COVFRAC unmatched-lands-in " +
          " ".join(f"{f}:{n}" for f, n in lost.most_common(6)))

    per_obj_n, per_obj_d = collections.Counter(), collections.Counter()
    for a, fn in sites.items():
        o = obj(fn)
        if o:
            per_obj_d[o] += 1
    for p in hit:
        o = obj(sites[p])
        if o:
            per_obj_n[o] += 1

    print("\nper kernel/bpf object:")
    for o, d in sorted(per_obj_d.items(), key=lambda kv: -per_obj_n.get(kv[0], 0)):
        n = per_obj_n.get(o, 0)
        if n or d >= 300:
            mark = '*' if o in VERIFICATION_PASS else ' '
            print(f" {mark} {o:22s} {n:6d} / {d:6d}  {100 * n / d:5.1f}%")

    vn = sum(per_obj_n.get(o, 0) for o in VERIFICATION_PASS)
    vd = sum(per_obj_d.get(o, 0) for o in VERIFICATION_PASS)

    per = collections.defaultdict(lambda: [0, 0])
    for a, fn in sites.items():
        if obj(fn) in VERIFICATION_PASS:
            per[fn][1] += 1
    for p in hit:
        fn = sites[p]
        if obj(fn) in VERIFICATION_PASS:
            per[fn][0] += 1
    never = sum(t for h, t in per.values() if h == 0)
    partial = sum(t - h for h, t in per.values() if h > 0)

    print(f"\nCOVFRAC verification_pass hit={vn} total={vd} pct={100 * vn / vd:.1f}")
    print(f"COVFRAC split reached={vn} unreached_in_entered_functions={partial} "
          f"in_functions_never_entered={never}")
    print(f"COVFRAC share_of_our_coverage_that_is_the_verifier="
          f"{100 * vn / max(1, len(hit)):.1f}")

    groups, rest = collections.Counter(), []
    for fn, (h, t) in per.items():
        if h:
            continue
        for name, pat in THEMES:
            if re.search(pat, fn):
                groups[name] += t
                break
        else:
            rest.append((t, fn))
    print(f"\nnever-entered, by theme ({never} points):")
    for name, n in groups.most_common():
        print(f"COVFRAC theme {name:30s} {n:5d}")
    print(f"COVFRAC theme {'everything else':30s} {sum(t for t, _ in rest):5d}")
    rest.sort(reverse=True)
    print("  largest ungrouped: " + ", ".join(f"{f}({t})" for t, f in rest[:10]))


main()

File: crates/pipeline/tests/coverage_fraction.rs
//! `seen_pcs` finally has a denominator (devlog 0086).
//!
//! Since 0078 the fuzz loop has reported a coverage number — 4789, 5203, 5406, 5443 — and
//! every one of them was a bare count. A count says the number went up. It never says how
//! much is left, and this project's standing rule is that a numerator without a
//! denominator is not evidence. `scripts/coverage-fraction.py` supplies the other half by
//! enumerating every `__sanitizer_cov_trace_pc` return address in vmlinux, which is
//! exactly the set of points KCOV can report.
//!
//! These tests pin the report, not the fuzzer. What they protect is the MEASUREMENT: if
//! the capture and the image ever come apart — a KASLR slide, a stale `.lab/build/vmlinux`
//! like the one 0085 found — the intersection collapses and every percentage in the report
//! becomes fiction while still reading like a result. That is the failure mode worth a
//! loud test.

const REPORT: &str = include_str!("fixtures/calibration/coverage-fraction.txt");

fn covfrac_field(key: &str) -> f64 {
    for line in REPORT.lines() {
        if let Some(rest) = line.strip_prefix("COVFRAC ") {
            for tok in rest.split_whitespace() {
                if let Some(v) = tok.strip_prefix(key) {
                    return v.parse().unwrap_or_else(|_| panic!("bad value for {key}: {tok}"));
                }
            }
        }
    }
    panic!("no COVFRAC field {key} in the report");
}

fn theme(name: &str) -> u64 {
    for line in REPORT.lines() {
        if let Some(rest) = line.strip_prefix("COVFRAC theme ") {
            let rest = rest.trim_end();
            if let Some(v) = rest.strip_prefix(name) {
                return v.trim().parse().unwrap_or_else(|_| panic!("bad theme line: {rest}"));
            }
        }
    }
    panic!("no theme {name}");
}

#[test]
fn the_capture_and_the_image_are_the_same_kernel() {
    // THE GUARD THAT MATTERS. 5370 PCs were recorded in the guest; 5234 of them land
    // exactly on an address this script located statically as a coverage point. That is
    // not something a wrong image can produce by accident — a KASLR slide or a different
    // build gives ~0% overlap, not 97%. Every other assertion in this file is downstream
    // of this one.
    let observed = covfrac_field("observed=");
    let matched = covfrac_field("matched=");
    assert!(observed > 5000.0, "the run must have recorded a real PC set: {observed}");
    assert!(
        matched / observed > 0.95,
        "capture and vmlinux disagree ({matched}/{observed}) — re-check which kernel \
         .lab/build/vmlinux actually is before believing any percentage below"
    );
}

#[test]
fn the_unmatched_remainder_is_accounted_for_and_not_one_sided() {
    // 136 PCs did not match. They are not a mystery and not a bias: a jump table inside
    // .text desynchronises the objdump walk, so the walk loses sites on BOTH sides of the
    // ratio. The report names where they land, and they land in functions we do reach.
    let unmatched = covfrac_field("unmatched=");
    let observed = covfrac_field("observed=");
    assert!(unmatched / observed < 0.05, "too much unexplained: {unmatched}/{observed}");
    let lands = REPORT
        .lines()
        .find(|l| l.starts_with("COVFRAC unmatched-lands-in"))
        .expect("the unmatched PCs must be attributed, not merely counted");
    assert!(
        lands.contains("check_mem_access") || lands.contains("do_check_common"),
        "unmatched PCs should fall inside functions we demonstrably enter: {lands}"
    );
}

#[test]
fn the_verification_pass_has_a_real_denominator_now() {
    let hit = covfrac_field("hit=");
    let total = covfrac_field("total=");
    let pct = covfrac_field("pct=");
    assert!(total > 5_000.0, "the denominator must be the whole pass, not a corner: {total}");
    // A fraction, in both directions: neither "we have seen nothing" nor a saturated 100%
    // that would mean the denominator was drawn too tight to be informative.
    assert!(pct > 5.0 && pct < 90.0, "not a usable fraction: {pct}");
    assert!((hit / total * 100.0 - pct).abs() < 0.2, "the report contradicts itself");
    // 0085 reported seen_pcs=5443 with no scale at all. This is the scale.
    assert_eq!(hit as u64, 3104);
    assert_eq!(total as u64, 10244);
}

#[test]
fn most_of_what_we_touch_is_actually_the_verifier() {
    // A coverage-guided loop can drift: `bpf_prog_load` drags in vmalloc, per-cpu
    // allocation and the JIT, and every new PC there counts as "coverage" while telling
    // us nothing about verification logic. 59% says the guidance is aimed at the right
    // half of the kernel; if this ever falls, the corpus is being rewarded for noise.
    assert!(
        covfrac_field("share_of_our_coverage_that_is_the_verifier=") > 50.0,
        "the coverage signal has drifted off the verifier"
    );
}

#[test]
fn the_two_kinds_of_unreached_are_separated() {
    // 33% sits in functions we have NEVER entered — a feature the grammar cannot express
    // at all — and 37% in blocks of functions we do enter. These are different problems
    // with different fixes, and a single "70% unreached" would hide that.
    let reached = covfrac_field("reached=");
    let partial = covfrac_field("unreached_in_entered_functions=");
    let never = covfrac_field("in_functions_never_entered=");
    assert!((reached + partial + never - covfrac_field("total=")).abs() < 0.5);
    assert!(never > 2_000.0 && partial > 2_000.0, "both halves are large: {never} {partial}");
}

#[test]
fn the_largest_unreached_region_is_named_and_it_is_kfuncs() {
    // THE AIM. 1318 of the 3397 never-entered points are kfunc and BTF-call handling —
    // 13% of the entire verification pass behind one feature the grammar has never
    // emitted. Everything else is smaller by a factor of four.
    //
    // When this test fails because kfuncs stopped being the largest theme, that is the
    // frontier moving and the number should be re-read, not the assertion relaxed.
    let kfunc = theme("kfunc / BTF calls");
    let others = [
        "log + disasm printing",
        "reference / spin lock",
        "attach target / trampoline",
        "callbacks / subprogs",
        "dynptr / iterator",
        "packet pointers",
    ];
    for o in others {
        assert!(kfunc > theme(o) * 2, "{o} has caught up with kfuncs — re-read the report");
    }
    assert!(kfunc > 1_000);
}

#[test]
fn the_numeric_core_is_the_part_we_have_saturated() {
    // The complement of the aim, and the reason the hunt has produced zero findings.
    // Our oracles are strongest on scalar arithmetic, and that is exactly where coverage
    // is nearly exhausted: cnum.o at 95%, tnum.o at 65%, states.o at 59%. The thin
    // regions are the ones our oracles say the least about — types, references, kfuncs.
    // The instrument is not starved of programs; it is pointed at a solved surface.
    let line = REPORT
        .lines()
        .find(|l| l.contains("cnum.o"))
        .expect("cnum.o must appear in the per-object table");
    let pct: f64 = line
        .split_whitespace()
        .last()
        .unwrap()
        .trim_end_matches('%')
        .parse()
        .unwrap();
    assert!(pct > 90.0, "the numeric core is no longer saturated: {line}");
}
